core/src/main/scala/com/pingcap/tispark/TiBatchWrite.scala

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2017 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pingcap.tispark
+
+import com.pingcap.tikv.TiBatchWriteUtils
+import com.pingcap.tikv.codec.KeyUtils
+import com.pingcap.tikv.key.{Key, RowKey}
+import com.pingcap.tikv.region.TiRegion
+import com.pingcap.tikv.row.ObjectRowImpl
+import com.pingcap.tikv.util.KeyRangeUtils
+import org.apache.spark.Partitioner
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.TiContext
+import org.slf4j.LoggerFactory
+
+/**
+ * An ugly implementation of batch write framework, which will be
+ * replaced by spark api.
+ */
+object TiBatchWrite {
+  private final val logger = LoggerFactory.getLogger(getClass.getName)
+
+  type SparkRow = org.apache.spark.sql.Row
+  type TiRow = com.pingcap.tikv.row.Row
+  type TiDataType = com.pingcap.tikv.types.DataType
+
+  def writeToTiDB(rdd: RDD[SparkRow], tableRef: TiTableReference, tiContext: TiContext) {
+    // TODO: region pre-split
+    // pending: https://internal.pingcap.net/jira/browse/TISPARK-69
+
+    // TODO: lock table
+    // pending: https://internal.pingcap.net/jira/browse/TIDB-1628
+
+    // shuffle data in same task which belong to same region
+    val shuffledRDD = {
+      val tiKVRowRDD = rdd.map(sparkRow2TiKVRow)
+      shuffleToSameRegion(tiKVRowRDD, tableRef, tiContext)
```

Comment:
s/shuffleToSameRegion/shuffleKeyToSameRegion

Reply:
will fix in next patch